model/delivery/AutoStartProctorService.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2022 (original work) Open Assessment Technologies SA;
+ *
+ */
+
+declare(strict_types=1);
+
+namespace oat\ltiProctoring\model\delivery;
+
+use common_session_Session as Session;
+use oat\oatbox\log\LoggerService;
+use oat\oatbox\user\User;
+use oat\taoDelivery\model\execution\DeliveryExecution;
+use oat\taoDelivery\model\execution\DeliveryExecutionInterface;
+use oat\taoLti\models\classes\LtiException;
+use oat\taoLti\models\classes\TaoLtiSession;
+use oat\taoLti\models\classes\user\LtiUser;
+use oat\taoProctoring\model\execution\DeliveryExecution as ProctoredDeliveryExecution;
+use oat\taoQtiTest\models\TestSessionService;
+
+class AutoStartProctorService
+{
+    private const CUSTOM_LTI_AUTOSTART = 'custom_autostart';
+
+    /**
+     * @var TestSessionService
+     */
+    private $testSessionService;
+
+    /**
+     * @var LoggerService
+     */
+    private $logger;
+
+    public function __construct(TestSessionService $testSessionService, LoggerService $logger)
+    {
+        $this->testSessionService = $testSessionService;
+        $this->logger = $logger;
+    }
+
+    public function execute(DeliveryExecution $deliveryExecution, User $user, Session $session): ?string
+    {
+        if (false === $this->validateAutoStart($deliveryExecution, $user)) {
+            return null;
+        }
+
+        $deliveryExecution->getImplementation()->setState(ProctoredDeliveryExecution::STATE_AUTHORIZED);
+
+        return $this->getUrlRunDeliveryExecution($deliveryExecution, $session);
+    }
+
+    private function validateAutoStart(DeliveryExecution $deliveryExecution, User $user): bool
+    {
+        if (null === $this->testSessionService->getTestSession($deliveryExecution)) {
```

Comment:
For me it would be clearer to have this condition extracted to a private method that explains that this is the way we check if the test has already been previously authorized and started